### PR TITLE
Merge PR: Export and import evm state (#717)

### PR DESCRIPTION
**Pull request by zhongqiuwood**
* add ut about ammswap liquidity with zero supply

* alpha model of exporting evm contracts into files

* add temp log

* add temp log

* add todo

* remove unused log

* optimize file operation

* the divorce of Storage

* optimze Storage read operation

* export txlogs into files

* fix the export panic

* optimize the filepath

* optimize the filepath

* finish tx logs read operation

* fix MustUnmarshalJSON problem

* fix the logtx path issue

* remove read&write operation into another file

* Unfinished: use sync to write data

* FINISH: write evm genesis data in sync

* change the limit number of goroutine pool

* add comment

* lower the goroutine pool limited

* add sync read when InitGenesis

* add set operation directly

* optimize the init Genesis.json

* change gosum

* add logs

* lower the goroutine pool limited

* move log position

* add comments

* go fmt

* tmp change

* Update go.mod

* fix evm ut race

* fix evm ut race

* Using the race detector when install

* add log (#716)

* add log

* bump github.com/okex/cosmos-sdk v0.39.2-okexchain8

* update message

Co-authored-by: Green <gl12138@live.com>

* Merge PR: optimize lock & path (#718)

* add log

* optimize path

* optimize path

* add initGoroutinePool

* change gosum

* bump to v0.16.8.1

Co-authored-by: Green <gl12138@live.com>
Co-authored-by: meng <805442788@qq.com>
Co-authored-by: evan.han <iloversevan@gmail.com>

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/okex/okexchain/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/okex/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/evm/export_operation.go [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+		storageCount++
+	}
+	iterator.Close()
+}
+
+func exportStorage(ctx sdk.Context, k Keeper, addr ethcmn.Address, db dbm.DB) {
+	defer wg.Done()
+
+	k.ForEachStorage(ctx, addr, func(key, value ethcmn.Hash) bool {
+		prefix := types.AddressStoragePrefix(addr)
```
Comment (banishee) [inline x/evm/export_operation.go]: it is better not to initialize the persistent prefix in a loop

File: x/evm/export_operation.go [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+		db.Set(append(prefix, key.Bytes()...), value.Bytes())
```
Comment (banishee) [inline x/evm/export_operation.go:165]: there is an unhandled error, or just ignore?

File: x/evm/export_operation.go [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+	// read Storage From file
+	addGoroutine()
+	go syncReadStorageFromFile(ctx, logger, k, address)
+}
+
+// exportToDB export EVM code and storage to leveldb
+func exportToDB(ctx sdk.Context, k Keeper, address ethcmn.Address, codeHash []byte) {
+	if code := k.GetCode(ctx, address); len(code) > 0 {
+		// TODO repeat code
+		evmByteCodeDB.Set(append(types.KeyPrefixCode, codeHash...), code)
```
Comment (banishee) [inline x/evm/export_operation.go]: there is an unhandled error, or just ignore?

File: x/evm/export_operation.go [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+		k.SetStateDirectly(ctx, address, ethcmn.BytesToHash(iterator.Key()[len(prefix):]), ethcmn.BytesToHash(iterator.Value()))
+		storageCount++
+	}
+	iterator.Close()
+}
+
+func exportStorage(ctx sdk.Context, k Keeper, addr ethcmn.Address, db dbm.DB) {
+	defer wg.Done()
+
+	k.ForEachStorage(ctx, addr, func(key, value ethcmn.Hash) bool {
```
Comment (banishee) [inline x/evm/export_operation.go]: there is an unhandled error, or just ignore?

File: x/evm/export_operation.go [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+
+// exportToDB export EVM code and storage to leveldb
+func exportToDB(ctx sdk.Context, k Keeper, address ethcmn.Address, codeHash []byte) {
+	if code := k.GetCode(ctx, address); len(code) > 0 {
+		// TODO repeat code
+		evmByteCodeDB.Set(append(types.KeyPrefixCode, codeHash...), code)
+		codeCount++
+	}
+
+	wg.Add(1)
```
Comment (banishee) [inline x/evm/export_operation.go]: should it limit the number of goroutine?

File: x/evm/genesis.go [hunk 6]
```diff
@@ -89,24 +114,42 @@ func ExportGenesis(ctx sdk.Context, k Keeper, ak types.AccountKeeper) GenesisSta
...
+				codeCount++
+			}
+			if storage, err = k.GetAccountStorage(ctx, addr); err != nil {
+				panic(err)
+			}
+			storageCount += uint64(len(storage))
+		case filesMode:
+			exportToFile(ctx, k, addr)
+		case dbMode:
+			exportToDB(ctx, k, addr, ethAccount.CodeHash)
```
Comment (KamiD) [inline x/evm/genesis.go:135]: need to return after call exportToDB

File: x/evm/genesis.go [hunk 7]
```diff
@@ -89,24 +114,42 @@ func ExportGenesis(ctx sdk.Context, k Keeper, ak types.AccountKeeper) GenesisSta
...
+			code = csdb.GetCode(addr)
+			if code != nil {
+				codeCount++
+			}
+			if storage, err = k.GetAccountStorage(ctx, addr); err != nil {
+				panic(err)
+			}
+			storageCount += uint64(len(storage))
+		case filesMode:
+			exportToFile(ctx, k, addr)
```
Comment (KamiD) [inline x/evm/genesis.go:133]: need to return after call exportToFile

File: x/evm/genesis.go [hunk 8]
```diff
@@ -44,23 +50,32 @@ func InitGenesis(ctx sdk.Context, k Keeper, accountKeeper types.AccountKeeper, d
...
+		case dbMode:
+			importFromDB(ctx, k, address, ethAcc.CodeHash)
+		default:
+			panic("unsupported import mode")
 		}
 	}
 
-	k.SetChainConfig(ctx, data.ChainConfig)
+	// wait for all data to be set into db
+	wg.Wait()
```
Comment (KamiD) [inline x/evm/genesis.go]: move Wait to case of filesMode

File: x/evm/export_operation.go [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+func syncWriteAccountCode(ctx sdk.Context, k Keeper, address ethcmn.Address) {
+	defer finishGoroutine()
+
+	code := k.GetCode(ctx, address)
+	if len(code) != 0 {
+		file := createFile(filepath.Join(codePath, address.String()+codeFileSuffix))
+		writer := bufio.NewWriter(file)
+		defer closeFile(writer, file)
+		writeOneLine(writer, hexutil.Bytes(code).String())
+		atomic.AddUint64(&codeCount, 1)
```
Comment (summerpro) [inline x/evm/export_operation.go:243]: It is best to use atomic.AddUint64 to prevent concurrency risks

File: x/evm/genesis.go [hunk 10]
```diff
@@ -89,24 +116,44 @@ func ExportGenesis(ctx sdk.Context, k Keeper, ak types.AccountKeeper) GenesisSta
...
+			Code:    code,
 			Storage: storage,
 		}
 
 		ethGenAccounts = append(ethGenAccounts, genAccount)
 		return false
 	})
+	// wait for all data to be written into files or db
+	if mode == filesMode || mode == dbMode {
+		wg.Wait()
```
Comment (KamiD) [inline x/evm/genesis.go:152]: dbMode has not init wg

Comment (ylsGit) [reply x/evm/genesis.go:152]: Not needed

File: x/evm/export_operation.go [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+	addGoroutine()
+	go syncReadStorageFromFile(ctx, logger, k, address)
+}
+
+// exportToDB export EVM code and storage to leveldb
+func exportToDB(ctx sdk.Context, k Keeper, address ethcmn.Address, codeHash []byte) {
+	if code := k.GetCode(ctx, address); len(code) > 0 {
+		// TODO repeat code
+		evmByteCodeDB.Set(append(types.KeyPrefixCode, codeHash...), code)
+		codeCount++
```
Comment (summerpro) [inline x/evm/export_operation.go:126]: It is best to use atomic.AddUint64 to prevent concurrency risks

Comment (ylsGit) [reply x/evm/export_operation.go:126]: There is no concurrency risk

**Review by banishee — DISMISSED** [review]

**Review by banishee — APPROVED** [review]

**Review by KamiD — APPROVED** [review]

**Review by xiangjianmeng — APPROVED** [review]

**Review by summerpro — APPROVED** [review]

**Pull request merged**